052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,147 @@
+
+[Kubernetes suggest](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts:~:text=It%20is%20strongly%20recommended%20for%20controllers%20to%20always%20%22force%22%20conflicts%2C%20since%20they%20might%20not%20be%20able%20to%20resolve%20or%20act%20on%20these%20conflicts.) using the above _force_ method to make sure Strimzi have control over the parts they want. As described in the Kubernetes docs _"This forces the operation to succeed, changes the value of the field, and removes the field from all other managers' entries in managedFields"_, this obviously means if another party is _forcing_ the same field that they will create a loop of ownership but this would be rare as operators should only care about their own fields.
+
+[Kubernetes recommend](https://kubernetes.io/docs/reference/using-api/server-side-apply/#upgrading-from-client-side-apply-to-server-side-apply:~:text=the%20object%20doesn%27t%20have%20to%20be%20read%20beforehand) omitting the `GET` call from the Kubernetes API, given that we only want to send the API our desired fields and leave it up to the API to solve.
+The thing here is that by removing `GET`, we would always need to `UPDATE` with the desired status and allow Kubernetes to resolve it. This would in theory be 1 API call made by Strimzi per resource per reconciliation as opposed to more. 
+[Kuberenetes suggest](https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/#reconstructive-controllers) there are two ways to write controllers with Server Side Apply, the first being a modification to the existing "Read-Modify-Update" method that Strimzi is currently using, to compute the changes in a similar fashion but only on the fields Strimzi owns. The second (and recommended) way is that you reconstruct the whole resource for every reconciliation (of each resource) and Apply that, letting Kubernetes manage it. The advantages and disadvantages are discussed in the above link. Notably a NO OP Apply is not very taxing on the API and can reduce the number of calls by quite a lot. **Ideally Strimzi would do it the second way, that is we can remove all comparison logic and storing of state and all GET requests, and just rebuild the resource every time and Server Side Apply it.**
+
+Existing Resources (Pods/ConfigMaps/etc) contain a patch strategy in the [upstream docs](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/), this explains how Server Side Apply will merge fields together, this will be unique to every field and if no strategy is listed then it can be assumed the whole field is changed each time. If this occurs and is not a desired behaviour then we can stick to the Read -> edit -> apply method instead for those examples with a justification.
```

Comment:
What does this part mean? `If this occurs and is not a desired behavior then we can stick to the Read -> edit -> apply method instead for those examples with a justification.`? I understand the first sentence as the Kube API spec defining how the fields will be merged. But I'm not sure I understand where would it not work, how would we understand it does not work and how will we implement the fallback you suggest.

Reply:
You had previously commented saying ConfigMaps might not work as desired with SSA (Server-Side Apply) and so I put this line as an alternative to using SSA, but I don't think it's a good idea as it means maintaining both SSA and the old (existing) method of apply.

Reply:
Yeah, I mentioned that. But it was just something I heard. I think we need to verify if it is really the case and if it is needed. If it would not be needed, it is definitely easier to avoid this.

Reply:
I shall test this locally and update :)

Reply:
I gave this a quick try. Good and bad news:
* It seems to delete the fields it owns. So from this perspective, it should work fine for us as we should be the ones owning the config maps / secrets and nobody should edit them. And if someone edits them, it is their problem to also remove the fields I guess. So we might not need this at least for Secrets and CMs (not sure why Pods were listed here as they are in general not editable, so not sure how it applies to them).
* To make this work, one needs to have a stable identity for the field manager owning the fields in the map which should be covered in the proposal. Please see the separate comment I starts for it.

Reply:
Nice - thanks for updating! That saves me a job today 😄 
Yes I am discussing the field manager more in the next update.